samza-core/src/main/java/org/apache/samza/zk/ZkBarrierListener.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.zk;
+
+/**
+ * An interface for listening to {@link ZkBarrierForVersionUpgrade} related events
+ */
+public interface ZkBarrierListener {
+  /**
+   * Invoked when the root of barrier for a given version is created in Zk
+   *
+   * @param version Version associated with the Barrier
+   */
+  void onBarrierCreated(String version);
+
+  /**
+   * Invoked when the data written to the Barrier state changes
+   *
+   * @param version Version associated with the Barrier
+   * @param state {@link org.apache.samza.zk.ZkBarrierForVersionUpgrade.State} value
+   */
+  void onBarrierStateChanged(String version, ZkBarrierForVersionUpgrade.State state);
+
+  /**
+   * Invoked when Barrier encounters error
+   *
+   * @param version Version associated with the Barrier
+   * @param t Throwable describing the cause of the barrier error
+   */
+  void onBarrierError(String version, Throwable t);
```

Comment:
Ah.. This method can be called when the barrier throws any exception. For now, we are not actually handling the errors from the barrier. For example, if there is a duplicate processor, "join" will throw an error saying ZkNodeExists. We don't handle it now, but potentially we should be able to handle it later on and cleanly shutdown the processor or handle intelligently. For now, I didn't want to include error handling in ZkBarrier. At least not in the same PR.